File: gearGauge/gearGauge/Models/Gear.swift
//
//  Gear.swift
//  gearGauge
//
//  Created by Shaun Hutchinson on 03/11/2025.
//

import SwiftUI
import SwiftData

/// A piece of fitness gear which is tracked by the user.
/// Stores distance travelled for a piece of gear, like shoes or a bicycle
@available(iOS 26, *)
@Model
final class Gear: BaseEntity {
    // MARK: main properties
    /// name of the gear
    var name: String
    
    /// The starting distance for the gear in kilometres.
    /// This allows users to set a starting value if gear was already used before tracking began
    var initialDistance: Double
    
    /// Cached total distance from all associated workouts in kilometres.
    /// This is denormalized for performance - updated when workouts are linked/unlinked.
    var cachedTotalWorkoutDistance: Double
    
    /// Maximum distance before replacement in kilometres
    var maxDistance: Double
    /// optional notes about the gear
    var notes: String?
    /// if the gear is the primary gear for the user
    var isPrimary: Bool
    /// if the gear is currently active
    var isActive: Bool
    /// gear start date
    var startDate: Date
    /// gear end date (retirement date)
    var endDate: Date?
    
    // MARK: Raw enum value handling
    
    /// what category the gear is (stored as raw String value for SwiftData)
    private var typeRawValue: String
    
    /// Raw values of workout types (stored for SwiftData compatibility)
    private var workoutTypeRawValues: [String]
    
    /// The type of gear (computed from raw value)
    /// SwiftData cannot store enums, just primitive values
    var type: GearType {
        get {
            GearType(rawValue: typeRawValue) ?? .shoes
        }
        set {
            typeRawValue = newValue.rawValue
        }
    }
    
    /// What types of workouts this gear is for (computed from raw values)
    var workoutTypes: [WorkoutType] {
        get {
            workoutTypeRawValues.compactMap { WorkoutType(rawValue: $0) }
        }
        set {
            workoutTypeRawValues = newValue.map { $0.rawValue }
        }
    }
    
    // MARK: Computed fields
    
    /// Current distance travelled by this gear in kilometres.
    /// Computed as initialDistance + cached workout distances for O(1) access.
    var currentDistance: Double {
        return initialDistance + cachedTotalWorkoutDistance
    }
    
    var currentDistanceMiles: Double { Double.ConvertToMi(currentDistance) }
    
    var maxDistanceMiles: Double { Double.ConvertToMi(maxDistance) }
    
    // MARK: - Relationships
    /// Workouts associated with this gear item
    /// When a workout is deleted, it's removed from this gear's workout list (nullify)
    /// The inverse relationship is defined on the Workout.gear property
    @Relationship(deleteRule: .nullify, inverse: \Workout.gear)
    var workouts: [Workout]?
    
    // MARK: - Initialisation
    init(
        name: String,
        type: GearType,
        initialDistance: Double = 0,
        maxDistance: Double,
        notes: String? = nil,
        isPrimary: Bool = false,
        isActive: Bool = true,
        startDate: Date,
        endDate: Date? = nil,
        workoutTypes: [WorkoutType] = [],
        workouts: [Workout] = []
    ) {
        self.name = name
        self.typeRawValue = type.rawValue
        self.initialDistance = initialDistance
        self.cachedTotalWorkoutDistance = 0 // Initialize to 0, will be recalculated
        self.maxDistance = maxDistance
        self.notes = notes
        self.isPrimary = isPrimary
        self.isActive = isActive
        self.workouts = workouts
        self.startDate = startDate
        self.endDate = endDate
        self.workoutTypeRawValues = workoutTypes.map { $0.rawValue }
        
        // Call parent initializer with defaults
        super.init()
    }
    
    // MARK: - Cache Management
    
    /// Recalculates and updates the cached total workout distance.
    /// This should be called when workouts are added, removed, or modified.
    /// Only counts workouts that are not soft-deleted.
    func recalculateCachedDistance() {
        cachedTotalWorkoutDistance = workouts?.filter { !$0.isDeleted }.reduce(0) { $0 + $1.totalDistance } ?? 0
    }
    
}

extension Gear {
    static func SampleGear() -> Gear {
        return self.init(
            name: "Asics Gel Kayano",
            type: .shoes,
            initialDistance: 900,
            maxDistance: 1000,
            notes: "Great running shoes",
            isPrimary: true,
            isActive: true,
            startDate: Date.newDateTime(year: 2025, month: 3, day: 15),
            workoutTypes: [.outdoorRun, .indoorRun],
            workouts: Workout.SampleWorkouts()
        )
    }
}

File: gear-gauge-rn/src/models/Gear.ts
import { randomUUID } from 'expo-crypto';
import type { GearType } from './GearType';
import type { WorkoutType } from './WorkoutType';
import { MAXIMUM_GEAR_DISTANCE } from './constants';

/**
 * A piece of fitness gear tracked by the user (e.g. running shoes, bicycle).
 * Mirrors the Swift `Gear` model in gearGauge/gearGauge/Models/Gear.swift
 */
export interface Gear {
  /** Unique identifier (UUID string). */
  id: string;
  /** User-facing name for this gear item. */
  name: string;
  /** Category of gear (shoes, bicycle, etc.). */
  type: GearType;
  /**
   * Starting distance in kilometres — for gear that was pre-used before
   * tracking began. Defaults to 0 for brand-new gear.
   */
  initialDistance: number;
  /** Maximum recommended distance in kilometres before replacement. */
  maxDistance: number;
  /** Whether this is the user's primary/default gear. */
  isPrimary: boolean;
  /** Whether this gear is currently active (not retired). */
  isActive: boolean;
  /** Date the gear was put into service (ISO-8601). */
  startDate: string;
  /** Retirement date if the gear has been retired (ISO-8601). */
  endDate?: string;
  /** Optional free-form notes about the gear. */
  notes?: string;
  /** WorkoutType values this gear is applicable to. */
  workoutTypeIds: WorkoutType[];
  /** ISO-8601 timestamp of when this record was created. */
  createdAt: string;
  /** ISO-8601 timestamp of when this record was last updated. */
  updatedAt: string;
  /** Monotonically-incrementing version for conflict resolution. */
  version: number;
  /** Soft-delete flag — true when the gear has been retired/removed. */
  isDeleted: boolean;
  /**
   * Current cumulative distance in kilometres.
   * Computed as `initialDistance + sum of all associated workout distances`.
   * Populated at runtime by the data layer or hooks — not a class getter.
   */
  currentDistance: number;
}

/**
 * Compute the current distance for a gear item given its initial distance
 * and the distances (in km) of all workouts associated with it.
 *
 * @example
 * computeCurrentDistance({ initialDistance: 100 }, [5, 10, 3.5])
 * // => 118.5
 */
export function computeCurrentDistance(
  gear: Pick<Gear, 'initialDistance'>,
  workoutDistances: number[],
): number {
  const totalWorkoutDistance = workoutDistances.reduce((sum, d) => sum + d, 0);
  return gear.initialDistance + totalWorkoutDistance;
}

// ── GearInput & factory ─────────────────────────────────────────────────────

/**
 * User-supplied fields for creating a new gear item.
 * Audit fields (`id`, `createdAt`, `updatedAt`, `version`, `isDeleted`) and the
 * computed `currentDistance` are filled in by {@link createGear}.
 */
export interface GearInput {
  name: string;
  type: GearType;
  /** Starting distance in km for pre-used gear. Defaults to 0. */
  initialDistance?: number;
  /** Maximum distance in km before replacement. */
  maxDistance: number;
  /** Optional free-form notes. */
  notes?: string;
  /** Whether this is the primary gear. Defaults to false. */
  isPrimary?: boolean;
  /** Whether the gear is active (not retired). Defaults to true. */
  isActive?: boolean;
  /** Date the gear was put into service (ISO-8601). Defaults to now. */
  startDate?: string;
  /** WorkoutType values this gear applies to. Defaults to []. */
  workoutTypeIds?: WorkoutType[];
}

/** Result of validating a {@link GearInput}. */
export interface GearValidationResult {
  valid: boolean;
  /** Field-name → human-readable error message. Empty when `valid` is true. */
  errors: Partial<Record<keyof GearInput, string>>;
}

/**
 * Create a fully-populated `Gear` entity from user input, applying defaults and
 * audit metadata. Mirrors the Swift `Gear.init` + `BaseEntity.init`.
 */
export function createGear(input: GearInput): Gear {
  const now = new Date().toISOString();
  const initialDistance = input.initialDistance ?? 0;

  return {
    id: randomUUID(),
    name: input.name,
    type: input.type,
    initialDistance,
    maxDistance: input.maxDistance,
    notes: input.notes,
    isPrimary: input.isPrimary ?? false,
    isActive: input.isActive ?? true,
    startDate: input.startDate ?? now,
    workoutTypeIds: input.workoutTypeIds ?? [],
    createdAt: now,
    updatedAt: now,
    version: 1,
    isDeleted: false,
    currentDistance: initialDistance,
  };
}

// ── Sorting ─────────────────────────────────────────────────────────────────

/**
 * Sort gear for display. Mirrors `GearViewModel.fetchAllGear`:
 * 1. primary gear first, then
 * 2. active (1) → inactive (2) → retired (3), then
 * 3. name ascending.
 */
export function sortGear(gear: Gear[]): Gear[] {
  return [...gear].sort((a, b) => {
    if (a.isPrimary !== b.isPrimary) {
      return a.isPrimary ? -1 : 1;
    }

    const rankA = gearRank(a);
    const rankB = gearRank(b);
    if (rankA !== rankB) {
      return rankA - rankB;
    }

    return a.name.localeCompare(b.name);
  });
}

/** Display rank: retired (3) < inactive (2) < active (1). */
function gearRank(gear: Gear): number {
  if (gear.endDate) return 3; // retired
  if (gear.isActive) return 1; // active
  return 2; // inactive
}

// ── Validation ──────────────────────────────────────────────────────────────

/**
 * Validate user input before creating/updating a gear item.
 * Rules: non-empty name; 0 < maxDistance ≤ MAXIMUM_GEAR_DISTANCE;
 * initialDistance ≥ 0.
 */
export function validateGearInput(input: GearInput): GearValidationResult {
  const errors: GearValidationResult['errors'] = {};

  if (!input.name || input.name.trim().length === 0) {
    errors.name = 'Name is required.';
  }

  if (
    input.maxDistance === undefined ||
    Number.isNaN(input.maxDistance) ||
    input.maxDistance <= 0
  ) {
    errors.maxDistance = 'Max distance must be greater than 0.';
  } else if (input.maxDistance > MAXIMUM_GEAR_DISTANCE) {
    errors.maxDistance = `Max distance cannot exceed ${MAXIMUM_GEAR_DISTANCE} km.`;
  }

  if (
    input.initialDistance !== undefined &&
    (Number.isNaN(input.initialDistance) || input.initialDistance < 0)
  ) {
    errors.initialDistance = 'Initial distance cannot be negative.';
  }

  return { valid: Object.keys(errors).length === 0, errors };
}

// ── Predicates ──────────────────────────────────────────────────────────────

/** True when the gear has been soft-deleted. */
export function isGearDeleted(gear: Gear): boolean {
  return gear.isDeleted;
}

/** True when the gear is active and not deleted. */
export function isGearActive(gear: Gear): boolean {
  return gear.isActive && !gear.isDeleted;
}

/** True when the gear is the primary gear and not deleted. */
export function isGearPrimary(gear: Gear): boolean {
  return gear.isPrimary && !gear.isDeleted;
}
